src/utils/normalize.ts

```diff
@@ -0,0 +1,94 @@
+import { add0x, isValidHexAddress } from '@metamask/utils';
+import type { Hex } from '@metamask/utils';
+import BN from 'bn.js';
+
+type EIP712Domain = {
+  verifyingContract: string;
+};
+
+type SignTypedMessageDataV3V4 = {
+  types: Record<string, unknown>;
+  domain: EIP712Domain;
+  primaryType: string;
+  message: unknown;
+};
+
+/**
+ * Normalizes the messageData for the eth_signTypedData
+ *
+ * @param messageData - The messageData to normalize.
+ * @returns The normalized messageData.
+ */
+export function normalizeTypedMessage(messageData: string) {
+  let data;
+  try {
+    data = parseTypedMessage(
+      messageData,
+    ) as unknown as SignTypedMessageDataV3V4;
+  } catch (e) {
+    // Ignore normalization errors and pass the message as is
+    return messageData;
+  }
+
+  const { verifyingContract } = data.domain ?? {};
+
+  if (!verifyingContract) {
+    return messageData;
+  }
+
+  data.domain.verifyingContract = normalizeContractAddress(verifyingContract);
+
+  return JSON.stringify(data);
+}
+
+/**
+ * Parses the messageData to obtain the data object for EIP712 normalization
+ *
+ * @param data - The messageData to parse.
+ * @returns The data object for EIP712 normalization.
+ */
+function parseTypedMessage(data: string) {
+  if (typeof data !== 'string') {
+    return data;
+  }
+
+  try {
+    return JSON.parse(data);
+  } catch (e) {
+    throw new Error(`Invalid message data for normalization. data: ${data}`);
+  }
+}
+
+/**
+ * Normalizes the address to a hexadecimal format
+ *
+ * @param address - The address to normalize.
+ * @returns The normalized address.
+ */
+function normalizeContractAddress(address: string): Hex {
+  const addressHex = address as Hex;
+  if (isValidHexAddress(addressHex)) {
```

Comment:
It appears that we need to use a type assertion because `isValidHexAddress` takes a `Hex`. However, we should be able to avoid the type assertion by using `isStrictHexString`, which narrows its argument to a `Hex` if given a hex string:

``` suggestion
  if (isStrictHexString(address) && isValidHexAddress(address)) {
```

That means we shouldn't need to use `addressHex` below, or upcast it to a `string`.

Reply:
Good call, done in https://github.com/MetaMask/eth-json-rpc-middleware/pull/309/commits/a8e8e716d05de8b8d46afff0d397dcda2cf977bc